### PR TITLE
Introduce snarkvm_nonnative

### PR DESCRIPTION
**Pull request by raychu86**
## Motivation

This PR introduces the `snarkvm_nonnative` module.

Currently missing `to_constraint_field` impls.

## Test Plan

Tests have been added for nonnative gadget arithmetic and byte/bit conversion.

### CODE REVIEW

File: utilities/src/biginteger/biginteger.rs [hunk 1]
```diff
@@ -100,6 +100,12 @@ pub trait BigInteger:
     /// leading zeros.
     fn to_bits(&self) -> Vec<bool>;
```
Comment (howardwu) [inline utilities/src/biginteger/biginteger.rs:101]: Can you rename all usages of this to `to_bits_be` and update the comment to be more descriptive?

File: nonnative/src/reduce.rs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+//     boolean::Boolean,
+//     eq::EqGadget,
+//     fields::{fp::FpVar, FieldVar},
+//     R1CSVar,
+// };
+// use ark_relations::{
+//     ns,
+//     r1cs::{ConstraintSystemRef, Result as R1CSResult},
+// };
+// use ark_std::{cmp::min, marker::PhantomData, vec, vec::Vec};
```
Comment (howardwu) [inline nonnative/src/reduce.rs:30]: This should be removed

File: nonnative/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+//! This library provides the non-native field gadget for the `snarkVM` constraint-writing platform.
+//! The non-native field gadget can be used as a standard `FieldVar`, given
+//! reasonable non-native gadget parameters.
+//!
+//! This file contains the implementation of three structs:
+//!
+//! - `NonNativeFieldParams` specifies the constraint prime field (called `BaseField`),
+//!     the simulated prime field (called `TargetField`), and internal parameters
+//!     searched by the Python script (see `README.md`).
```
Comment (raychu86) [inline nonnative/src/lib.rs:25]: This documentation refers to a README.md which we do not currently have for `snarkvm_nonnative`.

File: nonnative/src/allocated_nonnative_field_var.rs [hunk 4]
```diff
@@ -0,0 +1,982 @@
...
+
+    /// Compute the negate of a nonnative field element
+    pub fn negate<CS: ConstraintSystem<BaseField>>(&self, cs: &mut CS) -> Result<Self, SynthesisError> {
+        let zero = Self::zero(&mut cs.ns(|| "zero"))?;
+        zero.sub(&mut cs.ns(|| "sub"), self)
+    }
+
+    /// Compute the inverse of a nonnative field element
+    pub fn inverse<CS: ConstraintSystem<BaseField>>(&self, cs: &mut CS) -> Result<Self, SynthesisError> {
+        let inverse = Self::alloc_input(&mut cs.ns(|| "alloc_input"), || {
```
Comment (howardwu) [inline nonnative/src/allocated_nonnative_field_var.rs:329]: Should this be `alloc` instead of `alloc_input`?

File: nonnative/src/allocated_nonnative_field_var.rs [hunk 5]
```diff
@@ -0,0 +1,982 @@
...
+            &mut self_reduced,
+            &mut other_reduced,
+        )?;
+
+        let mut prod_limbs = Vec::new();
+
+        // We currently only support constraint optimization
+
+        for z_index in 0..2 * params.num_limbs - 1 {
+            prod_limbs.push(FpGadget::alloc_input(
```
Comment (howardwu) [inline nonnative/src/allocated_nonnative_field_var.rs:411]: Should this also be `alloc` instead of `alloc_input`?

File: nonnative/src/allocated_nonnative_field_var.rs (hunk 4 above)
Comment (raychu86) [reply nonnative/src/allocated_nonnative_field_var.rs:329]: Yep should be changed. The original notation was `new_witness` which should be equivalent to our `alloc`.

Comment (raychu86) [reply nonnative/src/allocated_nonnative_field_var.rs]: https://github.com/AleoHQ/snarkVM/pull/78/commits/a4f346c81761a60457a8d5d935a696950f39b6d1

File: nonnative/src/allocated_nonnative_field_var.rs (hunk 5 above)
Comment (raychu86) [reply nonnative/src/allocated_nonnative_field_var.rs]: https://github.com/AleoHQ/snarkVM/pull/78/commits/a4f346c81761a60457a8d5d935a696950f39b6d1

File: nonnative/src/allocated_nonnative_field_var.rs [hunk 6]
```diff
@@ -896,10 +893,9 @@ impl<TargetField: PrimeField, BaseField: PrimeField> AllocGadget<TargetField, Ba
         let mut limbs = Vec::new();
 
         for (i, limb) in elem_representations.iter().enumerate() {
-            limbs.push(FpGadget::<BaseField>::alloc_input(
-                cs.ns(|| format!("alloc_input_{}", i)),
-                || Ok(limb),
-            )?);
+            limbs.push(FpGadget::<BaseField>::alloc(cs.ns(|| format!("alloc_{}", i)), || {
```
Comment (howardwu) [inline nonnative/src/allocated_nonnative_field_var.rs:895]: Why did you change this to `alloc`?

**Review by howardwu — APPROVED** [review]
